### PR TITLE
Jv spring boot project hw 4

### PR DESCRIPTION
**Pull request by ihorvasylenko23**

### CODE REVIEW

**Review by ohatsenko — APPROVED** [review]

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Please resolve your PRs so only changes related to the task are here. After that request review again

File: pom.xml [hunk 1]
```diff
@@ -46,13 +46,27 @@
...
 		</dependency>
-
-
 		<dependency>
 			<groupId>mysql</groupId>
 			<artifactId>mysql-connector-java</artifactId>
 			<version>8.0.33</version>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
```
Comment (vovasalyha) [inline pom.xml]: It's better to prefer starters over concrete dependencies

File: src/main/java/mate/academy/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
 public class CreateBookRequestDto {
-
+    @Size(min = 1, max = 255)
+    @NotNull
     private String title;
+    @NotNull
+    @Size(min = 1, max = 255)
     private String author;
+    @NotNull
     private String isbn;
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dto/CreateBookRequestDto.java:21]: isbn has more validation rules

File: src/main/java/mate/academy/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,15 +1,26 @@
...
+    @NotNull
+    @Positive
     private BigDecimal price;
+    @NotNull
```
Comment (vovasalyha) [inline src/main/java/mate/academy/dto/CreateBookRequestDto.java]: if i'm not mistaken, by requirements, description and coverImage are nullable

File: src/main/java/mate/academy/repositiry/SpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.repositiry;
+
+import org.springframework.data.jpa.domain.Specification;
+
+public interface SpecificationProvider<T> {
+    String getKey();
+
+    Specification<T> getSpecification(String [] params);
```
Comment (vovasalyha) [inline src/main/java/mate/academy/repositiry/SpecificationProvider.java]:
```suggestion
    Specification<T> getSpecification(String[] params);

```

File: src/main/java/mate/academy/repositiry/book/spec/PriceSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+@Component
+public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "price";
+    }
+
+    public Specification<Book> getSpecification(String [] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("price").in(Arrays.stream(params).toArray());
```
Comment (vovasalyha) [inline src/main/java/mate/academy/repositiry/book/spec/PriceSpecificationProvider.java]: I'm not sure if it's very useful to search something by list of prices (usually it's price ranges)

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -39,4 +45,35 @@ public BookDto findById(Long id) {
                 .orElseThrow(() -> new EntityNotFoundException("Can`t find employee by id: " + id));
         return bookMapper.toDto(book);
     }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/mate/academy/service/impl/BookServiceImpl.java:50]:
```suggestion
    @Override
    @Transactional
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+databaseChangeLog:
+  - include: 
+      file: db/changelog/changes/01-create-books-table.yaml
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: empty line

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Let's improve it a bit

File: src/main/java/mate/academy/service/impl/BookServiceImpl.java (hunk 6 above)
Comment (nacenik) [reply src/main/java/mate/academy/service/impl/BookServiceImpl.java:50]: Not fixed

File: src/main/java/mate/academy/repositiry/book/spec/PriceSpecificationProvider.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "price";
+    }
+
+    public Specification<Book> getSpecification(Double minPrice, Double maxPrice) {
```
Comment (nacenik) [inline src/main/java/mate/academy/repositiry/book/spec/PriceSpecificationProvider.java]: is this a good idea to use double when the price is BigDecimal?

**Review by nacenik — APPROVED** [review]
GJ, But fix some mistake after

**Pull request merged**